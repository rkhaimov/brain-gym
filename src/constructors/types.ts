import { Either } from '../either';

export type TypeConstructor<TLeft, TRight> = (
  value: unknown
) => Either<TLeft, TRight>;
