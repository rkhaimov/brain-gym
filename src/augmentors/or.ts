import { TypeNode } from '../core';
import { validate } from '../validate';
import { refine } from './refine';
import { unknown } from '../types/unknown';

export const or = <TLeft, TRight>(
  left: TypeNode<TLeft>,
  right: TypeNode<TRight>
): TypeNode<TLeft | TRight> =>
  refine(unknown(), {
    validate: (value) => {
      const lerrors = validate(left, value as TLeft);

      if (lerrors.length === 0) {
        return [];
      }

      const rerrors = validate(right, value as TRight);

      if (rerrors.length === 0) {
        return [];
      }

      return [...lerrors, ...rerrors];
    },
    defaults: left.defaults,
  });
