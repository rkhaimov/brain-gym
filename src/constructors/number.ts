import { TypeConstructor } from './types';
import { Either } from '../either';

export const number = (): TypeConstructor<string, number> => (value) => {
  if (typeof value === 'number') {
    return Either.right(value);
  }

  return Either.left('is not a number');
};

