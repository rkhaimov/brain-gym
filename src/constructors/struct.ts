import { TypeConstructor } from './types';
import { Either } from '../either';

export const struct =
  <TStruct extends Struct>(sct: TStruct): StructTypeConstructor<TStruct> =>
  (_value) => {
    if (typeof _value !== 'object') {
      return Either.left('is not an object');
    }

    if (_value === null) {
      return Either.left('is null');
    }

    const validations = Object.entries(sct).map(([key, property]) =>
      toKeyErrorPair(property(get(_value, key)), key)
    );

    return Either.sequenceA(validations).leftMap(
      toErrorStruct
    ) as StructEither<TStruct>;
  };

const get = (object: object, key: string): unknown =>
  (object as Record<string, unknown>)[key];

const toKeyErrorPair = <TLeft, TRight>(
  either: Either<TLeft, TRight>,
  key: string
): Either<[string, TLeft], TRight> => {
  return either.leftMap((left) => [key, left]);
};

const toErrorStruct = <TLeft>(
  errors: [string, TLeft][]
): Record<string, TLeft> => {
  return Object.fromEntries(errors);
};

type Struct = Record<string, TypeConstructor<unknown, unknown>>;

type StructTypeConstructor<TStruct> = TypeConstructor<
  StructToError<TStruct> | string,
  StructToType<TStruct>
>;

type StructEither<TStruct> = Either<
  StructToError<TStruct> | string,
  StructToType<TStruct>
>;

type StructToType<TStruct> = TStruct extends TypeConstructor<
  unknown,
  infer RType
>
  ? RType
  : { [P in keyof TStruct]: StructToType<TStruct[P]> };

type StructToError<TStruct> = TStruct extends TypeConstructor<
  infer RError,
  unknown
>
  ? RError
  : { [P in keyof TStruct]: StructToError<TStruct[P]> };
