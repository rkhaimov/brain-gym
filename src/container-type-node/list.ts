import { ToType, TypeNode } from '../core';
import { fromErrorMessage, prependErrorPathWith, validate } from '../validate';
import {
  ContainerTypeNode,
  createContainerTypeNode,
} from './createContainerTypeNode';

export const list = <TChildren extends TypeNode>(
  tn: TChildren
): ContainerTypeNode<ToType<TChildren>[], TChildren> =>
  createContainerTypeNode({
    validate: (value) => {
      if (Array.isArray(value)) {
        return value.flatMap((element, index) =>
          validate(tn, element).map(prependErrorPathWith(index))
        );
      }

      return [fromErrorMessage('Not a list')];
    },
    defaults: () => [],
    children: () => tn,
  });
