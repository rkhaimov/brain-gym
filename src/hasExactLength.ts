import { TypeNode, TypeNodeOperator } from './core';
import { fromErrorMessage, fromKindAndPayload } from './validate';
import { validateConcatMap } from './validateConcatMap';

declare module 'validation-messages' {
  interface ValidationMessages {
    hasExactLength: number;
  }
}

export const hasExactLength =
  <TType extends { length: number }, TTypeNode extends TypeNode<TType>>(
    length: number
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.wrap(
      validateConcatMap((value) => {
        if (value.length === length) {
          return [];
        }

        return [fromKindAndPayload('hasExactLength', length)];
      })
    );
