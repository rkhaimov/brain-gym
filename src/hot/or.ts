import { TypeNode } from '../core';
import { validate } from '../validate';
import { unknown } from '../types/unknown';
import { refineMap } from '../operators/refineMap';
import { defaultsFrom } from '../operators/defaultsFrom';

export const or = <TLeft, TRight>(
  left: TypeNode<TLeft>,
  right: TypeNode<TRight>
) =>
  unknown().pipe(
    refineMap((value: TLeft | TRight) => {
      const lerrors = validate(left, value as TLeft);

      if (lerrors.length === 0) {
        return [];
      }

      const rerrors = validate(right, value as TRight);

      if (rerrors.length === 0) {
        return [];
      }

      return [...lerrors, ...rerrors];
    }),
    defaultsFrom(left)
  );
