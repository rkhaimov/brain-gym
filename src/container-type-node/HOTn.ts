import { TypeNode } from '../core';
import { translateFrom } from '../translate/translateFrom';
import { ContainerTypeNode } from './createContainerTypeNode';
import { list } from './list';

// TODO: Fix
// TODO: There are two types (Atomic and Compound) use it.
export const rewriteContainerWith = <TType, TChildren extends TypeNode>(
  tn: ContainerTypeNode<TType, TChildren>,
  rewrite: (tch: TChildren) => TChildren
): ContainerTypeNode<TType, TChildren> =>
  list(rewrite(tn.children())).wrap(
    translateFrom(tn)
  ) as unknown as ContainerTypeNode<TType, TChildren>;
