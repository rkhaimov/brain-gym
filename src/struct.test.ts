import { list } from './container-type-node/list';
import { hasExactLength } from './hasExactLength';
import { or } from './or';
import { rewrite } from './record-type-node/HOTn';
import { struct } from './record-type-node/struct';
import { translateTo } from './translateTo';
import { exact } from './type-node/types/exact';
import { number } from './type-node/types/number';
import { toFormValidationErrors, validate } from './validate';

describe('list', () => {
  describe('Validates and generates defaults', () => {
    it('validates with static checking', () => {
      const userTn = struct({ age: number(), weights: list(number()) });

      // @ts-expect-error // Static check
      expect(validate(userTn, '')).toMatchSnapshot();

      // @ts-expect-error // Static check
      expect(validate(userTn, {})).toMatchSnapshot();

      expect(
        validate(userTn, { age: 42, weights: [0, 1, '2'] })
      ).toMatchSnapshot();

      expectNoErrors(validate(userTn, { age: 42, weights: [0, 1, 2] }));
    });

    it('able to handle or validation', () => {
      const userTn = or(
        struct({
          type: exact('hr'),
          values: list(number()).wrap(
            hasExactLength(1),
            translateTo({
              hasExactLength: (size) =>
                `List has to have size of ${size} for HR`,
            })
          ),
        }),
        struct({
          type: exact('ecg'),
          values: list(number()).wrap(
            hasExactLength(3),
            translateTo({
              hasExactLength: (size) =>
                `List has to have size of ${size} for ECG`,
            })
          ),
        })
      ).wrap(translateTo({ exact: (type) => `Type must be exact ${type}` }));

      // @ts-expect-error // Static check
      expect(validate(userTn, {})).toMatchSnapshot();

      expect(
        validate(userTn, {
          // @ts-expect-error // Static check
          type: '',
        })
      ).toMatchSnapshot();

      expect(
        validate(userTn, {
          type: 'hr' as const,
          values: [],
        })
      ).toMatchSnapshot();

      expect(
        toFormValidationErrors(
          validate(userTn, {
            type: 'hr' as const,
            values: [1, 2, 3],
          })
        )
      ).toMatchSnapshot();

      expectNoErrors(
        validate(userTn, {
          type: 'hr' as const,
          values: [1],
        })
      );
    });

    it('allows to rewrite validation', () => {
      const tn = struct({
        age: number(),
        weight: number(),
      }).wrap(translateTo({ number: () => 'Not a number' }));

      const rtn = rewrite(tn, {
        age: (tn) =>
          tn.wrap(translateTo({ number: () => 'Age should be a number' })),
      });

      // @ts-expect-error static-check
      expect(validate(rtn, { age: '', weight: '' })).toMatchSnapshot();
    });
  });
});

function expectNoErrors(errors: unknown[]) {
  expect(errors.length).toBe(0);
}
