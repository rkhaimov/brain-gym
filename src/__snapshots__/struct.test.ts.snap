// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list Validates and generates defaults able to handle or validation 1`] = `
Array [
  Object {
    "kind": "exact",
    "message": "Type must be exact hr",
    "params": "hr",
    "paths": Array [
      "type",
    ],
  },
  Object {
    "kind": "unknown",
    "message": "Not a list",
    "params": undefined,
    "paths": Array [
      "values",
    ],
  },
  Object {
    "kind": "exact",
    "message": "Type must be exact ecg",
    "params": "ecg",
    "paths": Array [
      "type",
    ],
  },
  Object {
    "kind": "unknown",
    "message": "Not a list",
    "params": undefined,
    "paths": Array [
      "values",
    ],
  },
]
`;

exports[`list Validates and generates defaults able to handle or validation 2`] = `
Array [
  Object {
    "kind": "exact",
    "message": "Type must be exact hr",
    "params": "hr",
    "paths": Array [
      "type",
    ],
  },
  Object {
    "kind": "unknown",
    "message": "Not a list",
    "params": undefined,
    "paths": Array [
      "values",
    ],
  },
  Object {
    "kind": "exact",
    "message": "Type must be exact ecg",
    "params": "ecg",
    "paths": Array [
      "type",
    ],
  },
  Object {
    "kind": "unknown",
    "message": "Not a list",
    "params": undefined,
    "paths": Array [
      "values",
    ],
  },
]
`;

exports[`list Validates and generates defaults able to handle or validation 3`] = `
Array [
  Object {
    "kind": "hasExactLength",
    "message": "List has to have size of 1 for HR",
    "params": 1,
    "paths": Array [
      "values",
    ],
  },
  Object {
    "kind": "exact",
    "message": "Type must be exact ecg",
    "params": "ecg",
    "paths": Array [
      "type",
    ],
  },
  Object {
    "kind": "hasExactLength",
    "message": "List has to have size of 3 for ECG",
    "params": 3,
    "paths": Array [
      "values",
    ],
  },
]
`;

exports[`list Validates and generates defaults able to handle or validation 4`] = `
Array [
  Object {
    "message": "List has to have size of 1 for HR",
    "path": "values",
  },
  Object {
    "message": "Type must be exact ecg",
    "path": "type",
  },
]
`;

exports[`list Validates and generates defaults allows to rewrite validation 1`] = `
Array [
  Object {
    "kind": "number",
    "message": "Age should be a number",
    "params": undefined,
    "paths": Array [
      "age",
    ],
  },
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [
      "weight",
    ],
  },
]
`;

exports[`list Validates and generates defaults validates with static checking 1`] = `
Array [
  Object {
    "kind": "object",
    "message": "object: no params",
    "params": undefined,
    "paths": Array [],
  },
]
`;

exports[`list Validates and generates defaults validates with static checking 2`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [
      "age",
    ],
  },
  Object {
    "kind": "unknown",
    "message": "Not a list",
    "params": undefined,
    "paths": Array [
      "weights",
    ],
  },
]
`;

exports[`list Validates and generates defaults validates with static checking 3`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [
      "weights",
      2,
    ],
  },
]
`;
